Mathlib/LinearAlgebra/Lagrange.lean

```diff
@@ -620,6 +654,32 @@ theorem eval_interpolate_not_at_node' (hvs : Set.InjOn v s) (hs : s.Nonempty)
   simp only [mul_div_mul_left _ _ (eval_nodal_not_at_node hx), Pi.one_apply, mul_one]
 #align lagrange.eval_interpolate_not_at_node' Lagrange.eval_interpolate_not_at_node'
 
+/--
+The vanishing polynomial on a multiplicative subgroup is of the form X ^ n - 1
+-/
+theorem nodal_subgroup_eq_X_pow_card_sub_one (G : Subgroup (Units F)) [Fintype G] :
+  nodal G.carrier.toFinset (fun (x : Fˣ) => (x : F)) = X ^ (Fintype.card G) - C 1 := by
```

Comment:
```suggestion
/-- The vanishing polynomial on a multiplicative subgroup is of the form X ^ n - 1. -/
theorem nodal_subgroup_eq_X_pow_card_sub_one (G : Subgroup Fˣ) [Fintype G] :
  nodal (G : Set Fˣ).toFinset ((↑) : Fˣ → F) = X ^ (Fintype.card G) - 1 := by
```

Reply:
Also, I prefer `Finite` and `Nat.card` over `Fintype` and `Fintype.card` (since I usually care only _that_ something is finite rather than _how_) but the right choice depends on what pattern has been followed in nearby parts of the library.

In this case we could do:
```lean
attribute [local instance] Fintype.ofFinite

/-- The vanishing polynomial on a multiplicative subgroup is of the form X ^ n - 1. -/
theorem nodal_subgroup_eq_X_pow_card_sub_one (G : Subgroup Fˣ) [Finite G] :
  nodal (G : Set Fˣ).toFinset ((↑) : Fˣ → F) = X ^ (Nat.card G) - 1 := by
```
but I'll leave it to you to decide which is best here.

Reply:
You said this about #5738 too, both of these PRs come from the same project.

Reply:
All of this is meant for cryptographic applications, where the types are always finite and decidableEq. It's weird to me that there are like three different types of finiteness.

Reply:
I guess the point is that if you are doing a theorem, which has no data in it, we should use the weakest version of finiteness `Finite` so that every other version of finiteness can use it.

I feel sad that there seems to be a difference between `Nat.card` and `Fintype.card`, I am worried that I will have to prove lemmas later rewriting these.

Reply:
I don't really get the point of this local instance change. My understanding is it makes the lemma use the instance. If I am going to have `[Finite G]` but also use the `Fintype.ofFinite` instance, why not just use `Fintype`?

Reply:
`G` has to be cast to a `Finset` anyway, and I think `Finset` is the `Fintype` kind of finiteness, so it feels like this is the right notion here.

Reply:
Maybe what you are really advocating for is "we should use `Nat.card` everywhere we are currently using `Fintype.card`

I feel dumb, when I change the lemma to use `Nat.card` like you suggest, I have to fix the bugs. How to I fix the bugs? By putting `Nat.card_eq_fintype_card` everywhere.

Reply:
Maybe the real culprit is the entire fact that `Finset` uses the `Fintype` finiteness and not the `Finite` finiteness?

I feel like I could make an argument for "Lagrange.lean is a file where I would expect more computable notions of finiteness"

Reply:
Someone should write a docstring in `Mathlib.Data.Fintype.Card` about the difference between `Finite`/`Fintype`/`Fintype + DecidableEq`.

Reply:
Given that you need a `Finset` for `nodal`, and `Finset` wants a `Fintype`, there's a good case to leave this as you wrote it, using the data-bearing constructive finiteness (i.e., `Fintype`) rather than the better `Prop`-valued `Finite`. So let's just leave this as you wrote it.

Since I've brought this up before, I'll try to explain a bit: once you're talking about finiteness of subsets of a type there are two axes of finiteness giving a total of four possibilities:

1. Constructive, unbundled: `Fintype` + `Set.Finite`
2. Constructive, bundled: `Finset`
3. Non-constructive, unbundled: `Finite`
4. Non-constructive, bundled: missing

Of course it's ridiculous that such a simple concept has been made so complicated. The reason is that originally 1 & 2 were created and there is an excellent library of results about them but eventually their constructive nature becomes problematic. So we created 3. This is much better  but not used as much yet. I try to push it because it's the future. Maybe some hero will one day swap out `Fintype` for `Finite` in `Finset` and so turn 2 into 4.